smartsim/settingshold/common.py

```diff
@@ -0,0 +1,45 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import typing as t
+
+IntegerArgument = t.Dict[str, t.Optional[int]]
+StringArgument = t.Dict[str, t.Optional[str]]
+
+def process_env_vars(env_vars: StringArgument) -> None:
+    """Validate that user passed env vars are of correct type.
+    """
+    for key, value in env_vars.items():
+        if not isinstance(value, str) and value is not None:
+            raise TypeError(f"Value for '{key}' must be a string.")
+
+def process_args(launch_args: t.Dict[str, t.Union[str,int,float,None]]) -> None:
```

Comment:
I see! Does this just prevent errors that could come up down the line?

Reply:
It can! It can also make the method more flexible to re-use:

### Moves runtime errors to type-check time
This function takes a `launch_args` param, and we know (and want to guarantee to callers that) the parameter will not be mutated.

In the future some one were to edit this function, they may try to add this line:
```py
launch_args["some_arg_i_want_to_add_or_update"] = 123
```

If `launch_args` is a `dict`, this would be allowed because a dict can be read from and written to (i.e. it is invariant). This purposed line would be allow, but it would be an API break to the following snippet
```py
d = {}
process_args(d)
assert not d, "The docs say d should not mutated, so this a guaranteed to be true!"
```

If, on the other hand, `launch_args` is a `Mapping`, it is not allowed to be written into (i.e. it is covariant), and thus we can catch the API break at static analysis time!

### Reuse
Here is pretty safe, because the `launch_args` dict is composed of primitive keys/values, but if we had something like this:
```
class Animals:
    def speak(self):
        return "idk what to say :("

class Dog(Animal):
    def speak(self):
        return "woof"

class Cat(Animal):
    def speak(self):
        return "meow"

def print_animals_names_and_sounds(d: dict[str, Animal]):
    for k, v in d.items():
        print(k, "says", v.speak())

some_named_dogs = {
    "fido": Dog(), "lassie": Dog()}

print_animal_names_and_sounds(some_named_dogs) # <- This fails type checking!
                                                                                           #      But we know that it should work!
```
This is because `some_named_dogs` is inferred to be a `dict[str, Dog]`. But the type-checker does not know that what is happening in the impl of `print_animals_names_and_sounds`. I know that it takes a `dict[str, Animal]` and that `dict` are invariant. That means that this line COULD be present in the impl
```py
def print_animals_names_and_sounds(d: dict[str, Animal]):
    ...
    d["new_cat"] = Cat()
    ...
```
and if that were present then this happens:
```py
some_named_dogs = {  # infered to be `dict[str, Dog]`
    "fido": Dog(), "lassie": Dog()}
print_animal_names_and_sounds(some_named_dogs) # <- implicitly adds `"new_cat"` key
some_name_dogs["new_cat"]  # returns a `Cat`!
# WAIT A MINUTE! did that fn call implicitly change the type of an input parameter!!
```
which is why mypy flags that as an error!

`Mapping`s are not writable though, so it can be sure that the type of the input parameter will remain constant!